USACO Bronze - Social Distancing I (#5865)
cpinitiative/usaco-guide: solutions/bronze/usaco-1035.mdx

---
id: usaco-1035
source: USACO Bronze 2020 US Open
title: Social Distancing I
author: Aarav Sethi
---

> [Official Analysis](https://usaco.org/current/data/sol_socdist1_bronze_open20.html)

Reviewer: ```suggestion
[Official Analysis (C++)](https://usaco.org/current/data/sol_socdist1_bronze_open20.html)
```

## Implementation
**Time Complexity:** $\mathcal{O}(N\log N)$

<LanguageSection>

<CPPSection>
```cpp
#include <bits/stdc++.h>

using namespace std;

int compute_D(string stalls) {
	int d = INT32_MAX;
	int last1 = -1;

	for (int i = 0; i < stalls.length(); i++) {
		if (stalls[i] == '1') {
			if (last1 != -1) d = min(d, i - last1);
			last1 = i;
		}
	}

	return d;
}
pair<int, int> get_largest_zeros(string stalls) {
	// we can then get the two with maximum (end-start)
	pair<int, int> largestZero = {-1, -1};

	// get sizes
	int zeros_start = -1;
	for (int i = 0; i < stalls.size(); i++) {
		// if entered a '0' and not already counting, start counting
		if (stalls[i] == '0' && zeros_start == -1) { zeros_start = i; }

		// hit a '1' & we have counted some zeros, or we have reached the end
		if ((stalls[i] == '1' && zeros_start != -1) || i == stalls.size() - 1) {
			// insert the zero sequence
			pair<int, int> zero = {zeros_start, i};
			if ((zero.second - zero.first) > (largestZero.second - largestZero.first)) {
				largestZero = zero;
			}
			zeros_start = -1;
		}
	}

	return largestZero;
}

int main() {
	int n;
	cin >> n;
	string stalls;
	cin >> stalls;

> 	auto largest = get_largest_zeros(stalls);

Author: I tried that but it leads to WA for a couple of testcases, because sometimes we need to:
1. Fill the center of the largest
2. Recompute for the next largest
3. Fill the center of the next largest

and having it use the initial first largest, and second largest is not the same.

3. `00001000` - Place at the two far edges.
4. `10000000`/`00000001` - Place a cow in one of the far edges, and one in the largest gap.

We can now test all of these cases, and maximize $D$.

## Implementation
**Time Complexity:** $\mathcal{O}(N\log N)$

<LanguageSection>

<CPPSection>
```cpp
#include <bits/stdc++.h>

using namespace std;

int compute_D(string stalls) {
	int d = INT32_MAX;
	int last1 = -1;

	for (int i = 0; i < stalls.length(); i++) {
		if (stalls[i] == '1') {
			if (last1 != -1) d = min(d, i - last1);
			last1 = i;
		}
	}

	return d;
}
vector<pair<int, int>> get_zeros(string stalls) {
	// [(start, end)], we can then sort to maximize end-start
	vector<pair<int, int>> zeros;

	int zeros_start = -1;
	for (int i = 0; i < stalls.size(); i++) {
		// if entered a '0' and not already counting, start counting
		if (stalls[i] == '0' && zeros_start == -1) { zeros_start = i; }

		// hit a '1' & we have counted some zeros, or we have reached the end
		if ((stalls[i] == '1' && zeros_start != -1) || i == stalls.size() - 1) {
			// insert the zero sequence
			zeros.push_back({zeros_start, i});
			zeros_start = -1;
		}
	}

> 	sort(zeros.begin(), zeros.end(),

Reviewer: sorting w/ custom comparator hasn't been introduced yet

we only need two longest intervals, no? i think edi approach of finding location of longest interval and essentially splitting the longest interval twice is cleaner

		string c = stalls;

		for (int i = 0; i < 2; i++) {
			auto sZ = get_zeros(c);
			auto loc = sZ[0];
			int pos = (loc.first + loc.second) / 2;  // midpoint

			c[pos] = '1';
		}

		d = max(d, compute_D(c));
	}
	if (zeros.size() >= 1) {
		// 100000001
		// place one at 1/3 and one at 2/3
		auto loc = zeros[0];
		int gap = loc.second - loc.first;

		string c = stalls;
		c[loc.first + gap / 3] = '1';
		c[loc.first + (gap * 2) / 3] = '1';

		d = max(d, compute_D(c));
	}
	if (stalls[0] == '0' && stalls[stalls.size() - 1] == '0') {
		// 000010000
		// place at two edges
		string c = stalls;
		c[0] = '1';
		c[stalls.size() - 1] = '1';

		d = max(d, compute_D(c));
	}
	if (stalls[0] == '0') {
		// 00010001
		// place at furthest point & within largest zero
		string c = stalls;
		c[0] = '1';

		auto nZ = get_zeros(c);
		c[(nZ[0].first + nZ[0].second) / 2] = '1';

		d = max(d, compute_D(c));
	}
	if (stalls[stalls.size() - 1] == '0') {
		// 10010000
		// place at furthest point & within largest zero
		string c = stalls;
		c[stalls.size() - 1] = '1';

		auto nZ = get_zeros(c);
		c[(nZ[0].first + nZ[0].second) / 2] = '1';

		d = max(d, compute_D(c));
	}

> 	cout << d;

Reviewer: ```suggestion
	cout << d << '\n';
```

just a minor thing

---
id: usaco-1035
source: USACO Bronze 2020 US Open
title: Social Distancing I
author: Aarav Sethi
---

[Official Analysis](https://usaco.org/current/data/sol_socdist1_bronze_open20.html)

## Explanation
### Goal
- We are given a binary string with `0` representing an empty stall, and `1` representing an occupied stall.
- We need to find the two most optimal places to put cows, to maximize $D$.

> - $D$ is the minimum distance between two occupied stalls.

Reviewer: ```suggestion
- We need to find the two most optimal places to put cows, to maximize the minimum distance between two occupied stalls $D$.
```

IMO you don't need to restate the problem tho, since the statement is alr relatively clear/it's not that much of a reinterpretation

Reviewer: prolly, always nice to be a bit more concise

The sample testcase can be solved by fitting two cows in the two largest gaps.
**ex.** `10001001000010` -> `10`**`1`**`010010`**`1`**`0010`.

> But many more cases exist, for `1000001`, or `0001000` for example.

Reviewer: or something to that effect
